### ISSUE TITLE
Extending computed class name identifiers throws:

### ISSUE THREAD

**Issue by mohsen1** [google/closure-compiler#2182]
Input:

```js
var a = {
  foo: class {}
}

class bar extends a['foo']{
  
}
```

CC throws:

```
JSC_DYNAMIC_EXTENDS_TYPE: The class in an extends clause must be a qualified name. at line 6 character 18
class bar extends a['foo']{
                  ^
```

Is this by design? The JS code is valid.

**Comment by mohsen1** [comment]
I think CC should not treat bracket property accessor with string value as "dynamic". In my example `class bar extends a.foo` works but `class bar extends a['foo']` doesn't while both are the same.

TypeScript compiler can follow the code:

<img width="493" alt="screen shot 2016-12-04 at 10 40 08 am" src="https://cloud.githubusercontent.com/assets/543633/20868341/13efba00-ba0e-11e6-8f94-b7a8ad9057d9.png">

**Comment by MatrixFrog** [comment]
Can you elaborate a bit on what your use case is? i.e. why you don't want
to write 'a.foo' instead?

**Comment by mohsen1** [comment]
Webpack outputs code like this:

```js
class APIClientError extends __WEBPACK_IMPORTED_MODULE_3_helpers__["e" /* CustomError */] {
```

And I am using CC to minify Webpack output.

**Comment by alexeagle** [comment]
Duplicate of https://github.com/google/closure-compiler/issues/2106 ?

**Comment by ChadKillingsworth** [comment]
Not a duplicate - it's different. Fixing this would be a pretty simple change to the transpilation. Should be non-controversial as well.

**Comment by tomccabe** [comment]
Just ran into this as well trying to replace the ever-bloating Babel with CC. Sadly I'm getting otherwise great results, but this error pretty much prevents us from adopting CC.

**Comment by TheLarkInn** [comment]
> Webpack outputs code like this:

`class APIClientError extends __WEBPACK_IMPORTED_MODULE_3_helpers__["e" /* CustomError */] {`

@mohsen1 webpack would only replace that if you are extending a namespace / import / etc. that you have imported. Can you please share the original source?

**Comment by mohsen1** [comment]
I don't exactly remember the original code anymore but this is close enough:

```ts
// Errors.ts
export class CustomError {
// ....
}
```

```ts
// SomeFile.ts
import { CustomError } from './Errors.ts';

class APIClientError extends CustomError {
// ...
}
```

**Comment by TheLarkInn** [comment]
Thanks, @mohsen1 https://github.com/webpack/webpack/blob/master/lib/dependencies/HarmonyImportSpecifierDependency.js#L87-L99 if you wanted to hack.

**Comment by sokra** [comment]
As workaround:

``` js
// SomeFile.ts
import { CustomError } from './Errors.ts';

const CustomErrorTemp = CustomError;

class APIClientError extends CustomErrorTemp {
// ...
}
```

**Comment by vanhoavn** [comment]
The problem is from cases like this:

```javascript
class APIClientError extends CustomError<API> {
// ...
}
```

we cant do:

```javascript
const CustomErrorTemp = CustomError<API>;
```

**Comment by mohsen1** [comment]
You should be able to do that with `type`:

```ts
type CustomErrorTemp = CustomError<API>;

class APIClientError extends CustomErrorTemp {
// ...
}
```

**Comment by vanhoavn** [comment]
Hmm error:

```
error TS2693: 'CustomErrorTemp' only refers to a type, but is being used as a value here.
```

**Comment by mohsen1** [comment]
oh right... sorry for misleading comment

**Comment by Treeki** [comment]
I ran into this issue and decided to take a stab at fixing it (despite never having touched the Closure Compiler codebase before...). As far as I can tell, the most elegant solution is to allow CC to consider terms like `a.b['c']` as a qualified name, similar (but not equivalent!) to `a.b.c`.

I don't want to submit a PR for this as my fix isn't perfect -- it fails some tests -- but perhaps somebody more familiar with the codebase can build upon it and complete it: https://github.com/Treeki/closure-compiler/tree/allow-constant-getelem-in-qualified-names

**Comment by nichtverstehen** [comment]
I deal with this limitation by manually extracting the base into a variable. I wonder if applying such a transform mechanically could be a correct resolution for this issue.

I.e. for the code in the first comment, the rewriting could be:
```
var a = {
  foo: class {}
}

var barBase0 = a['foo'];
class bar extends barBase0 {
  // ...
}
```

Will this be typed as expected?

**Comment by paulfisher53** [comment]
I'm also experiencing this issue with webpack + closure compiler. Is there any ETA on a fix?

**Comment by ChadKillingsworth** [comment]
See #2997

**Comment by ChadKillingsworth** [comment]
Closed by #2997
